File: services/money/LifeOS.Money.Api/Domain/UnitDimension.cs
namespace LifeOS.Money.Api.Domain;

/// The physical dimension of a quantity on a spending or estimate line (ADR-0036). A line
/// carries a magnitude (<see cref="Line.Quantity"/>) plus one of these dimensions; the
/// display unit SYMBOL (kg/lb/L/gal/m/ft) is a pure client rendering of (dimension × the
/// owner's <see cref="UnitSystem"/>) and is never stored or converted by Money.
/// <see cref="Pieces"/> renders no symbol (a bare count); the others render one symbol
/// chosen by the client from the active unit system. Serialized as its underlying integer
/// on the wire (STJ default), like the service's other directly-embedded enums.
public enum UnitDimension {
    Pieces,
    Mass,
    Volume,
    Length,
}

File: apps/wallet/lib/features/money/domain/unit_dimension.dart
import 'unit_system.dart';

/// A generic quantity dimension on a line or want (Money ADR-0036). The backend
/// stores only the dimension — **never** a concrete unit (kg, lb, …). The symbol
/// is a UI-only cosmetic rendered from the dimension × the user's [UnitSystem]
/// preference (see [symbol]); **no conversions ever**.
///
/// Rides the wire as an int (STJ default), matching the Money service.
enum UnitDimension {
  pieces,
  mass,
  volume,
  length;

  /// Parses the wire int. Returns `null` for a missing/null value.
  static UnitDimension? fromWire(Object? value) {
    if (value is int) {
      return switch (value) {
        0 => UnitDimension.pieces,
        1 => UnitDimension.mass,
        2 => UnitDimension.volume,
        3 => UnitDimension.length,
        _ => null,
      };
    }
    return null;
  }

  /// The wire int the backend stores.
  int get wire => index;

  /// The cosmetic symbol for this dimension under [system], or `null` when the
  /// dimension has no symbol (Pieces is a bare count — `×2`, never `×2 pcs`).
  /// A finance annotation, not a measurement: switching [UnitSystem] relabels
  /// every quantity without touching its number.
  String? symbol(UnitSystem system) => switch (this) {
        UnitDimension.pieces => null,
        UnitDimension.mass => system == UnitSystem.metric ? 'kg' : 'lb',
        UnitDimension.volume => system == UnitSystem.metric ? 'L' : 'gal',
        UnitDimension.length => system == UnitSystem.metric ? 'm' : 'ft',
      };
}
